[Feature] replace dgl PRNG with pcg32

mfbalin opened the pull request:
## Description
A way of addressing #4785. Don't have to merge it like this, but can be just merged as well. I don't know if the existing tests rely on the underlying PRNG to pass.

## Changes
Replaced std::default_random_engine with pcg32, passed thread_id as the stream argument of pcg32.

I can close this PR if replacing the PRNG is not on the radar.

frozenbugs commented:
Hi @mfbalin , thanks very much for your help. In general, as the discussion goes in https://github.com/dmlc/dgl/issues/4785, we don't really see a good reason to add another external dependency for dgl code.
But I do agree we can switch to MT19937, which is reportedly better than minstd_rand0, and it is provided by std.

Would you mind to switch to that in this PR?
Thanks.

mfbalin commented:
> Hi @mfbalin , thanks very much for your help.
> 
> As the discussion goes in #4785. I agree pcg could be better than MT19937 based on the evidence, but we don't really see a good reason to add another external dependency for dgl code. I do agree we can switch to MT19937, which is reportedly better than minstd_rand0, and it is provided by std.
> 
> Would you mind to switch to that in this PR? Thanks.

Hi, the external dependency for pcg has already been added as part of #4668. If it wants to be removed, then a short implementation header containing pcg can be added and the submodule dependency could be removed. Taking this into account, if DGL still wants to switch to mt19937, I could change the PR.

frozenbugs commented:
> > Hi @mfbalin , thanks very much for your help.
> > As the discussion goes in #4785. I agree pcg could be better than MT19937 based on the evidence, but we don't really see a good reason to add another external dependency for dgl code. I do agree we can switch to MT19937, which is reportedly better than minstd_rand0, and it is provided by std.
> > Would you mind to switch to that in this PR? Thanks.
> 
> Hi @frozenbugs, the external dependency for pcg has already been added as part of #4668. If it wants to be removed, then a short implementation header containing pcg can be added and the submodule dependency could be removed. Taking this into account, if DGL still wants to switch to mt19937, I could change the PR. Or if the external dependency wants to be removed for sure, I could cook up a short pcg implementation by looking at the official code available. It takes only 20 lines to implement.

Yes, please move to mt19937, and remove the dependency in #4668 if possible. If you have a strong reason to use pcg (e.g. measurable performance improvement on DGL), we can revisit the decision. Since no-one here is an expert of pcg, I'd prefer to not have the pcg implemented since it increase the maintenance cost.

mfbalin commented:
I have removed the submodule required for #4668 by reimplementing pcg32 in "include/dgl/pcg32.h". I have also replaced the std::default_generator to std::mt19937_64 for RandomEngine. However, the new implementation of pcg32 should also be good to go, I checked and the code compiles when I replace std::default_generator with the implementation I added for pcg32. #4668 seems to just work with new implementation also.

mfbalin commented:
Because of the way I am using the PRNG in #4668, I can't use mt19937 instead. That is why I had to reimplement pcg32 to remove the submodule dependency because it is required.

frozenbugs commented on src/array/cpu/labor_pick.h:
```diff
@@ -24,6 +24,7 @@
 
 #include <dgl/array.h>
 #include <dgl/random.h>
+#include <dgl/pcg32.h>
```
Is it possible to use mt19937 in this code?

mfbalin replied:
It would make it much slower, and results wouldn't be theoretically guaranteed to be correct either. That code doesn't use the PRNG in the usual way, which is to seed it once and get numbers. What I do is to keep reseeding before getting a new number.

frozenbugs replied:
Thanks for educating me on this, do we have any measured results to show how much it is slower?
Also, why a different prng could make it "not theoretically guaranteed to be correct"? Conceptually they are just simulating random number.

If indeed we see the performance different, yes, we should switch to pcg.
Can you revert back to your original change, which is using pcg_random.hpp in include/dgl/random.h. And I am wondering whether you can include include/dgl/random.h in src/array/cpu/labor_pick.h, instead of including pcg32.h / pcg_random.hpp.

Again, thank you for your great and in-depth contribution.

mfbalin replied:
Usually, PRNG's are designed so that when you seed them once, every random number you sample after are independently identically distributed. But in my case, I require that the random numbers one gets by seeding with a different seed, every first random number you sample are independently identically distributed.

mfbalin replied:
In labor_pick.h, replacing
`pcg32(random_seed, t)` line with
```
std::seed_seq seq{(uint32_t)(random_seed >> 32), (uint32_t)random_seed, (uint32_t)(t >> 32), (uint32_t)t};
std::mt19937 ng(seq);
```
Causes 1000x slowdown.

mfbalin replied:
Can't even inspect if results are correct because the code is not terminating, as in comparing to the provided numbers in the original issue #4785.

mfbalin replied:
The reason why it is so slow is that `sizeof(std::mt19937)` returns 5000, so each time it needs to seed, it has to reinitialize a space of kilobytes as opposed to just two numbers as in pcg32.

mfbalin replied:
Also, I haven't run the benchmarks for it but I am certainly sure that the neighbor sampling cpu kernel will become at least 2x slower if RandomEngine switches from std::default_random_engine to std::mt19937/_64.

mfbalin replied:
One still needs to run benchmarks to see how much slower NeighborSampler will become when std::default_random_engine is swapped for pcg32.

frozenbugs commented on include/dgl/pcg32.h:
```diff
@@ -0,0 +1,72 @@
+/*!
+ *   Copyright 2014 Melissa O'Neill <oneill@pcg-random.org>
```
Is it legitimate to just copy like this?

mfbalin replied:
I didn't directly copy. I actually basically ported it to c++ but the operator()() is mostly the same, because that is the definition of pcg32 basically. Anybody who implements pcg32 implements operator()() in the same way. For the copyright, I didn't know if I should copyright it for myself or directly use the original author's, decided to do the latter.

mfbalin replied:
You can check the referenced file to see what I am talking about, https://github.com/imneme/pcg-c-basic/blob/master/pcg_basic.c

mfbalin commented:
I don't know why but requesting review again removed the other existing reviewer, it was not my intention to do so.

frozenbugs commented:
@dgl-bot

frozenbugs commented:
@dgl-bot

frozenbugs commented:
@dgl-bot
